File: LeetCode/Medium/3020_Find_The_Maximum_Number_Of_Elements_In_Subset/solution.py
"""
LeetCode 3020
Find the Maximum Number of Elements in Subset

Difficulty: Medium

Approach:
Count frequencies with a hash map. For each base x > 1, build the chain
x -> x^2 -> x^4 -> ... requiring freq >= 2 at each step except the center.
Handle x = 1 separately (odd count of 1s). Return the max chain length.

Time Complexity: O(n log log M)
Space Complexity: O(n)
"""

from typing import List
from collections import Counter


class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        count = Counter(nums)
        result = 1

        if 1 in count:
            ones = count[1]
            result = ones if ones % 2 == 1 else ones - 1

        for base in count:
            if base == 1:
                continue
            chain_length = 0
            current = base
            while current in count and count[current] >= 2:
                chain_length += 2
                current = current * current
            if current in count:
                chain_length += 1
            else:
                chain_length -= 1
            result = max(result, chain_length)

        return result

File: LeetCode/Medium/3020_Find_The_Maximum_Number_Of_Elements_In_Subset/solution.java
/*
LeetCode 3020
Find the Maximum Number of Elements in Subset

Difficulty: Medium

Approach:
Count frequencies with a hash map. For each base x > 1, build the chain
x -> x^2 -> x^4 -> ... requiring freq >= 2 at each step except the center.
Handle x = 1 separately (odd count of 1s). Return the max chain length.

Time Complexity: O(n log log M)
Space Complexity: O(n)
*/

import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maximumLength(int[] nums) {
        Map<Long, Integer> count = new HashMap<>();
        for (int num : nums) {
            count.merge((long) num, 1, Integer::sum);
        }

        int result = 1;

        if (count.containsKey(1L)) {
            int ones = count.get(1L);
            result = (ones % 2 == 1) ? ones : ones - 1;
        }

        for (long base : count.keySet()) {
            if (base == 1) continue;
            int chainLength = 0;
            long current = base;
            while (count.containsKey(current) && count.get(current) >= 2) {
                chainLength += 2;
                current = current * current;
            }
            if (count.containsKey(current)) {
                chainLength += 1;
            } else {
                chainLength -= 1;
            }
            result = Math.max(result, chainLength);
        }

        return result;
    }
}
